File: python-importer/oggwrap.py
"""把设备吐的裸 OPUS 定长包封成 Ogg/Opus。

CB08 与 QS668 同一种码流：40 字节 = 一个 20ms 帧（config 9，16kHz 宽带）。
Ogg 的 granule 走 48kHz 时钟，所以每包推进 960。
源自项目内 qs668-raw-opus-to-ogg.py，改成可导入模块。
"""
from __future__ import annotations

import struct

PACKET_LEN = 40
FRAME_MS = 20
GRANULE_PER_PACKET = 960          # 20ms @ 48kHz
_CRC_POLY = 0x04C11DB7


def _build_crc_table() -> list[int]:
    table = []
    for value in range(256):
        reg = value << 24
        for _ in range(8):
            reg = ((reg << 1) ^ _CRC_POLY) & 0xFFFFFFFF if reg & 0x80000000 else (reg << 1) & 0xFFFFFFFF
        table.append(reg)
    return table


_CRC_TABLE = _build_crc_table()


def _ogg_crc(data: bytes) -> int:
    crc = 0
    for byte in data:
        crc = ((crc << 8) & 0xFFFFFFFF) ^ _CRC_TABLE[((crc >> 24) & 0xFF) ^ byte]
    return crc


def _page(payloads: list[bytes], granule: int, serial: int, seq: int, flags: int) -> bytes:
    laces: list[int] = []
    for payload in payloads:
        remaining = len(payload)
        while remaining >= 255:
            laces.append(255)
            remaining -= 255
        laces.append(remaining)
    header = bytearray(b"OggS" + bytes([0, flags]))
    header += struct.pack("<QIII", granule, serial, seq, 0)
    header += bytes([len(laces)]) + bytes(laces)
    page = header + b"".join(payloads)
    page[22:26] = struct.pack("<I", _ogg_crc(page))
    return bytes(page)


def wrap(raw: bytes, sample_rate: int = 16000, tag: str = "CB08") -> bytes:
    """裸包 -> Ogg/Opus 字节。长度不是 40 的整数倍时截掉尾部残包并照常封装。"""
    usable = len(raw) - (len(raw) % PACKET_LEN)
    packets = [raw[i:i + PACKET_LEN] for i in range(0, usable, PACKET_LEN)]
    if not packets:
        raise ValueError("没有完整的 OPUS 包")

    serial, seq, pages = 0x51533638, 0, []
    head = (b"OpusHead" + bytes([1, 1]) + struct.pack("<H", 312)
            + struct.pack("<I", sample_rate) + struct.pack("<h", 0) + bytes([0]))
    tag_b = tag.encode()
    tags = b"OpusTags" + struct.pack("<I", len(tag_b)) + tag_b + struct.pack("<I", 0)
    pages.append(_page([head], 0, serial, seq, 0x02)); seq += 1
    pages.append(_page([tags], 0, serial, seq, 0x00)); seq += 1

    granule = 0
    for start in range(0, len(packets), 50):
        group = packets[start:start + 50]
        granule += GRANULE_PER_PACKET * len(group)
        last = start + 50 >= len(packets)
        pages.append(_page(group, granule, serial, seq, 0x04 if last else 0x00))
        seq += 1
    return b"".join(pages)


def duration_seconds(raw_len: int) -> float:
    return raw_len // PACKET_LEN * FRAME_MS / 1000


def looks_raw(data: bytes) -> bool:
    return len(data) >= PACKET_LEN and len(data) % PACKET_LEN == 0 and data[:4] != b"OggS"

File: Sources/LuyinbiCore/OggWrap.swift
import Foundation

/// 把设备吐的裸 OPUS 定长包封成 Ogg/Opus。
///
/// CB08 与 QS668 同一种码流：40 字节 = 一个 20ms 帧（config 9，16kHz 宽带）。
/// Ogg 的 granule 走 48kHz 时钟，所以每包推进 960。
public enum OggWrap {
    public static let packetLen = 40
    public static let frameMs = 20
    public static let granulePerPacket: UInt64 = 960     // 20ms @ 48kHz
    private static let crcPoly: UInt32 = 0x04C1_1DB7

    private static let crcTable: [UInt32] = {
        (0..<256).map { value -> UInt32 in
            var reg = UInt32(value) << 24
            for _ in 0..<8 {
                reg = (reg & 0x8000_0000) != 0 ? (reg << 1) ^ crcPoly : reg << 1
            }
            return reg
        }
    }()

    private static func oggCRC(_ data: [UInt8]) -> UInt32 {
        var crc: UInt32 = 0
        for byte in data {
            crc = (crc << 8) ^ crcTable[Int(((crc >> 24) & 0xFF) ^ UInt32(byte))]
        }
        return crc
    }

    private static func page(_ payloads: [[UInt8]], granule: UInt64,
                             serial: UInt32, seq: UInt32, flags: UInt8) -> [UInt8] {
        var laces: [UInt8] = []
        for payload in payloads {
            var remaining = payload.count
            while remaining >= 255 { laces.append(255); remaining -= 255 }
            laces.append(UInt8(remaining))
        }
        var header: [UInt8] = Array("OggS".utf8) + [0, flags]
        header += (0..<8).map { UInt8((granule >> ($0 * 8)) & 0xFF) }
        header += (0..<4).map { UInt8((serial >> ($0 * 8)) & 0xFF) }
        header += (0..<4).map { UInt8((seq >> ($0 * 8)) & 0xFF) }
        header += [0, 0, 0, 0]                       // CRC 占位
        header += [UInt8(laces.count)] + laces
        var p = header + payloads.flatMap { $0 }
        let crc = oggCRC(p)
        for i in 0..<4 { p[22 + i] = UInt8((crc >> (i * 8)) & 0xFF) }
        return p
    }

    public enum WrapError: Error { case noCompletePacket }

    /// 裸包 → Ogg/Opus。长度不是 40 的整数倍时截掉尾部残包。
    public static func wrap(_ raw: [UInt8], sampleRate: UInt32 = 16000,
                            tag: String = "CB08") throws -> [UInt8] {
        let usable = raw.count - (raw.count % packetLen)
        guard usable >= packetLen else { throw WrapError.noCompletePacket }
        let packets = stride(from: 0, to: usable, by: packetLen).map {
            Array(raw[$0..<($0 + packetLen)])
        }

        let serial: UInt32 = 0x5153_3638
        var seq: UInt32 = 0
        var pages: [[UInt8]] = []

        var head: [UInt8] = Array("OpusHead".utf8) + [1, 1]
        head += [UInt8(312 & 0xFF), UInt8(312 >> 8)]                 // pre-skip, LE
        head += (0..<4).map { UInt8((sampleRate >> ($0 * 8)) & 0xFF) }
        head += [0, 0, 0]                                            // output gain(2) + mapping(1)
        let tagBytes = Array(tag.utf8)
        var tags: [UInt8] = Array("OpusTags".utf8)
        tags += (0..<4).map { UInt8((UInt32(tagBytes.count) >> ($0 * 8)) & 0xFF) }
        tags += tagBytes + [0, 0, 0, 0]

        pages.append(page([head], granule: 0, serial: serial, seq: seq, flags: 0x02)); seq += 1
        pages.append(page([tags], granule: 0, serial: serial, seq: seq, flags: 0x00)); seq += 1

        var granule: UInt64 = 0
        for start in stride(from: 0, to: packets.count, by: 50) {
            let group = Array(packets[start..<min(start + 50, packets.count)])
            granule += granulePerPacket * UInt64(group.count)
            let last = start + 50 >= packets.count
            pages.append(page(group, granule: granule, serial: serial, seq: seq,
                              flags: last ? 0x04 : 0x00))
            seq += 1
        }
        return pages.flatMap { $0 }
    }

    public static func durationSeconds(rawLength: Int) -> Double {
        Double(rawLength / packetLen * frameMs) / 1000
    }

    public static func looksRaw(_ data: [UInt8]) -> Bool {
        data.count >= packetLen && data.count % packetLen == 0
            && Array(data.prefix(4)) != Array("OggS".utf8)
    }
}
